Update deployment addresses

jubeira opened the pull request:
# Description

Update deployment addresses and add step to deployment process to keep it up to date.
The changes in the PR were generated with `yarn build-address-lookup`.

## Type of change

- [ ] Bug fix <!-- (non-breaking change which fixes an issue) -->
- [ ] New feature <!-- (non-breaking change which adds functionality) -->
- [ ] Breaking change <!-- (would cause existing functionality to not work as expected) -->
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [x] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- N/A Complex code has been commented, including external interfaces
- N/A Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

## Issue Resolution

Closes #2445.

EndymionJkb commented:
So this already worked? We just need to run that existing task after deployments?

jubeira commented:
> So this already worked? We just need to run that existing task after deployments?

Yes, that should be pretty much it. This doesn't consider deprecated contracts (although as per off-thread chat it should be fine right?)

EndymionJkb commented:
Maybe also have it print a "green check" message when it succeeds, like the others? It works, but doesn't print anything.

nventuro commented:
We don't have any CI checks that look at the correctness of these files, do we?

EndymionJkb commented:
> No. Maybe we could add a task that generates them, and compares the generated files against the existing ones. That task will fail if these files are not updated during a new deployment. WDYT?

The other directories work that way, and the reason for resurrecting this was integrations/maxis wanted to rely on it to produce documentation, so we should have some guarantee that it's accurate (and complete).

jubeira commented:
Alright, I've modified the scripts a bit so that the address lookup file is generated all at once, and always respecting the same order. Before these changes, the file's entries could be ordered differently according to the order of the tasks added to it, which is not ideal.

I've also added the missing entries, and a CI task that checks whether the committed file is up to date as discussed. PTAL.

EndymionJkb commented on pkg/deployments/hardhat.config.ts:
```diff
@@ -256,24 +256,48 @@ task('get-action-ids-info', `Reconstructs all the permissions from TheGraph AP a
+  const addressLookupFileOk = checkContractDeploymentAddresses(tasks, hre.network.name);
+  if (!addressLookupFileOk) {
+    throw new Error(
+      `Address lookup file is incorrect for network ${hre.network.name}. Please re-generate file by running 'build-address-lookup' task`
```
```suggestion
      `Address lookup file is incorrect for network ${hre.network.name}. Please run 'build-address-lookup' to regenerate it`
```

EndymionJkb commented on pkg/deployments/src/network.ts:
```diff
@@ -42,25 +42,53 @@ export function getContractDeploymentTransactionHash(deployedAddress: string, ne
-export function saveContractDeploymentAddresses(task: Task): void {
-  if (task.network === 'hardhat') return;
+/**
+ * Saves file with canonical deployment addresses for all the given tasks in the given network,
```
```suggestion
 * Saves a file with the canonical deployment addresses for all tasks in a given network.
```

EndymionJkb commented on pkg/deployments/src/network.ts:
```diff
@@ -42,25 +42,53 @@ export function getContractDeploymentTransactionHash(deployedAddress: string, ne
+}
+
+/**
+ * Builds object that maps deployment addresses to {task ID, contract name} for all the given tasks.
```
```suggestion
 * Builds an object that maps deployment addresses to {task ID, contract name} for all given tasks.
```

EndymionJkb commented on pkg/deployments/src/network.ts:
```diff
@@ -42,25 +42,53 @@ export function getContractDeploymentTransactionHash(deployedAddress: string, ne
-    Object.entries(task.output({ ensure: false })).map(([name, address]) => [address, { task: task.id, name }])
-  );
+/**
+ * Returns true if the existing deployment addresses file stored in `CONTRACT_ADDRESSES_DIRECTORY` matches the
+ * canonical one for the given set of tasks in the given network; false otherwise.
```
```suggestion
 * Returns true if the existing deployment addresses file stored in `CONTRACT_ADDRESSES_DIRECTORY` matches the
 * canonical one for the given network; false otherwise.
```

EndymionJkb reviewed (approved):
One possible refinement. Macs tend to have lots of .DS_Store files strewn about from the Finder, and this causes all the tasks without --id flags to break, since they get included in the /tasks directory listing. (You have to know to delete the `tasks/.DS_Store` file.)

Consider filtering out hidden files along with the task file, to be Mac-friendly?
.filter((taskId) => !taskId.startsWith('00000000-') && !(taskId[0] == '.'))

Maybe even factor out that filtering code somehow, as it's in multiple places, though that might be hard given the structure of these files.

jubeira commented:
The address registry is happy, CI is happy, and Tim Cook is happy :+1:
